### ISSUE TITLE
No "fixed" parameter for Heading, TabBar

### ISSUE THREAD

**Issue by jhpedemonte** [maqetta/maqetta#856]
**Heading** and **TabBar** OAM files don't specify the `fixed` property, which is used when those widgets are placed within a **ScrollableView**.

Also, the existing docs aren't that great at pointing out this property.  For example, in the [Heading docs](http://docs.dojocampus.org/dojox/mobile/Heading), `fixed` is mentioned in the opening paragraph, but not in the **Constructor Params** section.  In the [TabBar docs](http://docs.dojocampus.org/dojox/mobile/TabBar), `fixed` is only mentioned once, in an example.

Not sure if any other widgets can also use `fixed`...

@JonFerraiolo @peller

---

See my comment from [Sept 26](https://github.com/maqetta/maqetta/issues/856#issuecomment-2205913) for use case scenarios.

**Comment by jhpedemonte** [comment]
So, following [OpenAjax Metadata Spec](http://www.openajax.org/member/wiki/OpenAjax_Metadata_1.0_Specification_Datatypes), I think this should be represented thusly:

``` xml
   <property name="fixed" datatype="fixedEnum" defaultValue="" />
   ...
   <enum name="fixedEnum" datatype="string">
      <option value=""/>
      <option value="top"/>
      <option value="bottom"/>
   </enum>
```

Jon, does that seem right?

Of course, we don't currently support `enum` in OAM. But using `enum` will allow us (in the future) to change the Properties view to show drop down menus for some properties, rather than only a text box.

**Comment by JonFerraiolo** [comment]
Yes, looks like the right way to go.

**Comment by jhpedemonte** [comment]
I just added the `fixed` prop to Heading and TabBar, as a simple string datatype.  I'll open a different bug for handling the enumeration.

Is `fixed` prop used by other widgets, like IconContainer?

**Comment by jhpedemonte** [comment]
Created #863 to track handling of enumerated property values.

**Comment by JonFerraiolo** [comment]
Transferred remainder of bug to me. The metadata is there, but visual editor does not support attribute properly yet. I'm working on it.

**Comment by JonFerraiolo** [comment]
I did a bunch of research on this and have concluded it's too risky to try to add support for "fixed" for M3, so I'm pushing this bug to M4, and for M3 I'm removing the "fixed" property from the OAM files.

Here are some notes:
- So far, it appears that "fixed" only applies to ScrollableView and not to View or SwapView. (I'm 80% certain)
- For ScrollableView, if there are child widgets with fixed="top" and/or fixed="bottom", ScrollableView rearranges the children such that the children with fixed="bottom" are moved to the top of the list of children, then children with fixed="top", and then all of the other children are placed inside of a DIV with class "ScrollableViewContainer". If no children have fixed="top" or fixed="bottom", then the DIV with class "ScrollableViewContainer" is not created. If DIV with class ScrollableViewContainer is present, then Dojo set DijitWidget.containerNode to that DIV. The Maqetta page editor right now assumes that all children widget are direct descendents of DijitWiget.containerNode. As a result, this dynamic rearrangement approach totally confuses the page editor. I tried to create a getChildren() helper function, and it generally solved this problem.
- ScrollableView doesn't store the "fixed" property on the DijitWidget object. Therefore, getPropertyValue() doesn't find that property. I tried to create a getPropertyValue() helper function, and it generally solved this problem.
- The killer problem, though, is that ScrollableView does some of its essential layout processing only at page parsing/startup time. Even the resizeAll() function in dojox.mobile won't result in a complete recalculation of layout because only at page parse time does ScrollableView figure out whether there are widgets with fixed="top" or fixed="bottom". Because of this, I tried adding some helper functions that would trigger a setSource() operation (i.e., reparse/rebuilt the page from the current model), and that generally caused things to appear in the correct locations, but this process seems to cause other parts of the page editor to have stale handles to widgets (and probably context and other things), causing various exceptions.
- I noticed lots of problems with our code along the way, one of which is that when we open an HTML file, we first call the Dojo parser, and then only at the very end, we create the silhouette and resize the IFRAME. But as I mentioned earlier, ScrollableView does some of its layout processing at parse time. I think what we need to do is create the silhouette and size the IFRAME _before_ calling the Dojo parser. However, this might cause other problems. As I was hacking, I tried more of a bandaid approach and added an extra setSource() call in setDevice(), which seemed to work, but means that on page load we would be parsing the HTML file twice.
- I had to split getChildren() helper into 2 separate calls: findChildren and getChildren. findChildren is called before getWidget() creates widgets as a side-effect (which of course isn't the best possible approach)
- I had to force property "disableFlashScrollBar" to true on ScrollableView because triggering layout calculations triggers scrollbar manipulation logic which can raise a null exception

**Comment by jhpedemonte** [comment]
Jon, you stated that you created helper functions for `getChildren` and `getPropertyValue`. Do you still have that code?

**Comment by jhpedemonte** [comment]
## Use cases when adding widget with "fixed" property:
1. **Adding widget to ScrollableView**
   - _Options_:
     - Smart input which immediately asks for **Label** and **fixed** (none, top, bottom). Then adding widget to ScrollableView should go through ScrollableView.addChild(), which does the proper thing for "fixed" widgets.
2. **Setting `fixed` prop on existing child widget of ScrollableView**
   - Currently does nothing, since widget is already inside of the `mblScrollableViewContainer` DIV.
     - _Options_:
       - Destroy existing widget, create new widget with same props plus the `fixed` prop, then add as child, which should call `ScrollableView.addChild()`.
       - After setting `fixed` property in source, for a refresh of the VE from source, which should result in widget being properly placed.
   - _See note below on handling property values_
   - ==> Fixed by 418fadff5b7064b651398765b30b5d2d103fc069.
3. **Moving widget with 'fixed' prop into ScrollableView**
   - ~~I assume this should work well, since it should go through `ScrollableView.addChild()`, which does the proper thing for "fixed" widgets.~~ **Edit**: Nope, this doesn't work like that. Need to find out why.
   - ==> Fixed by commits culminating in bde4041e794f3da0b72fcb3792645790700e151b.
4. **Remove 'fixed' prop from existing widget**
   - Options:
     - Remove `fixed` prop and call `ScrollableView.reparent()`.
       - Unfortunately, if it was `fixed="top"`, and there are other children widgets in ScrollableView, then the widget gets added to the bottom of the View, after all the other widgets -- which probably isn't what the user expects.
     - After removing fixed prop in source, refresh VE.
   - _See note below on handling property values_
   - ==> Fixed by 418fadff5b7064b651398765b30b5d2d103fc069.

---

**Handling property values in Dijit Widgets (and possible in all widget types)** _(added Oct 14)_
- When setting value in Properties view, the property and value are added as an HTML attribute in the Model but added as an Object property within context of the Visual Editor. The property is not added as a DOM attribute in the VE.
- When opening a file with a fixed header/footer (that is, an element has `fixed="top"` or `fixed="bottom"`), that value is not reflected in the Properties view; it shows no value for the `fixed` property.

==> Fixed by commit 9e7c12033bff064eaa35fd5c0168d684111cd6ab (Oct 26)

**Comment by jhpedemonte** [comment]
Ugh.  So, [Use Case 3](https://github.com/maqetta/maqetta/issues/856#issuecomment-2205913) (moving widget with 'fixed' prop into ScrollableView) is failing due to the `! this.acceptsHTMLChildren` check in `davinci.ve.DijitWidget.addChild()`.

That check was added since many of the Dojox Mobile containers (such as ScrollableView) accept both Dojo-based widgets and HTML elements, yet inherit from `dijit._Container`. The latter class is meant for containers which only take Dijit widgets as children, not HTML children. This is an inconsistency in between Dijit and Dojox Mobile. See #741 for more details.

In _Use Case 3_, when adding the Heading to ScrollableView, the latter widget has `acceptsHTMLChildren` set to `true`, so it goes down the "generic" (non-Dojo) route and doesn't call `ScrollableView.addChild()`, which is supposed to look for the `fixed` property and handle this Heading widget in a special case.

@peller, this is killing me.  It's really bad that the Dojox Mobile devs are ignoring the non-HTML-children constraints of `dijit._Container`.  I'm stumped as to how to handle this particular issue without regressing #514 and #741.

**Comment by jhpedemonte** [comment]
There is no easy way to call `_Container.addChild()` on a parent widget that takes both Dojo and HTML widgets and have it function as we want.  The reason is that `addChild()` needs to know where to place the new child and does so by calling `getChildren()`.  This function, however, only returns children that are Dijit widgets -- meaning that calling `addChild()` on parent widget with both Dijit and HTML widgets will result in widgets being placed in the incorrect location (see #741).

The only idea that comes to mind (and this is an ugly idea) is to override `dojo.place()` (called by `_Container.addChild()`). We could then specify the correct relative DOM node (could be node from either Dijit or HTML widget) and insertion index.

**Comment by JonFerraiolo** [comment]
Overriding dojo.place() would not be a good thing to do. Almost like overriding DOM's appendChild.

I don't know the details of the problems you are facing, but I wouldn't think container widgets such as ContentPane that take both Dojo and HTML widgets would use addChild(), and instead just use DOM methods (insertChild*, appendChild). In other words, instead of trying to get addChild() to work in this scenario by patching dojo.place(), maybe avoiding it is better.

**Comment by jhpedemonte** [comment]
Well, that's not an option.  ScrollableView does some funky things in its `addChild` method and needs to be called when placing a widget inside it.  Other Dijit/Dojox widgets may do the same.

I don't think the idea of hooking `dojo.place()` is that bad.  It would be something like this:
1. hook dojo.place
2. call Dijit's addChild()
3. unhook dojo.place

It's ugly, but it should work.  If you have a better idea, please let me know.

**Comment by jhpedemonte** [comment]
That last commit is on a branch; hasn't been checked in to master. It implements what I proposed [earlier](https://github.com/maqetta/maqetta/issues/856#issuecomment-2214683). Let me know what you think.

There's a big comment there that should explain everything.  If it hasn't, let me know and I'll update it.  Also, notice that `_addChildHooked` only gets called for DijitWidgets that implement `addChild()` and set `acceptsHTMLChildren`.

**Comment by jhpedemonte** [comment]
Next issue I'm running into with [Use Case 3](https://github.com/maqetta/maqetta/issues/856#issuecomment-2205913): when moving the Heading widget, the old widget gets destroyed and a new one created.  Not sure why that is.

This is what the widget DOM looks like before the move:

``` html
<h1 label=​"Fixed Heading" dojotype=​"dojox.mobile.Heading" fixed=​"top" id=​"h1_0" class=​"mblHeading mblFixedHeaderBar mblHeadingCenterTitle" style widgetid=​"h1_0">
   <span class="mblHeadingSpanTitle">Fixed Heading</span>
   <div class="mblHeadingDivTitle">Fixed Heading</div>
​</h1>​
```

And here's what it looks like after the move, when `ScrollableView.addChild()` is called:

``` html
<h1 label=​"Heading" id="Heading_0" dojotype=​"dojox.mobile.Heading" class=​"mblHeading" style widgetid=​"Heading_0">
   <span class="mblHeadingSpanTitle">Fixed Heading</span>
   <div class="mblHeadingDivTitle">Fixed Heading</div>
​</h1>​
```

As you can see, `label`, `id`, `class` attributes have changed, and `fixed` is no longer there.  The latter causes this to break, since ScrollableView no longer sees this as a fixed header and handles it differently.

Any ideas as to why we handle widget moves like this?

**Comment by jhpedemonte** [comment]
Actually, that last commit is silly.  The issue is with Dijit's `getChildren()` function, not `dojo.place()`.  I should just temporarily replace the parent widget's `getChildren()` with Maqetta's similar function; that should work.  Otherwise, commit is the same.

**Comment by jhpedemonte** [comment]
Moved [widget moving issue](https://github.com/maqetta/maqetta/issues/856#issuecomment-2218874) to #1056.

**Comment by jhpedemonte** [comment]
First, I made some changes to the [use cases comment](https://github.com/maqetta/maqetta/issues/856#issuecomment-2205913).

Second, it looks like the best way to fix the remaining cases is to **refresh the VE from source** whenever the `fixed` property is set.

After trying some things along those lines, I think I've found the perfect solution:  add support for a **metadata property** which would say that if a given **property is changed**, then **refresh from source**.  This would circumvent the normal process for property changes.

My proposal:
1. Add support for the `refreshSource` metadata property, which is a _boolean_ with default value of `false`.
2. Since this is a Maqetta specific property, we can't add it to the widget's OAM file.  Instead, we support it in the widget descriptor in the **widgets.json** file, but adding a top level `property` prop [1], which acts as a mixin for the `property` prop in the OAM file.  For example:
   
   ``` javascript
   {
       "name": "Heading",
       ...
       "property": {
           "fixed": {
               "refreshSource": true
           }
       }
       ...
   },
   ```
3. When a property with `refreshSource` set to true is changed, we refresh the VE from source, similar to how it is currently done when editing in the source editor.

<br>

[1] Unfortunately, this means that we could both `property` and `properties` in the widget descriptor.  Maybe rename `properties` to something else?

**Comment by JonFerraiolo** [comment]
Regarding putting refreshSource in the metadata, seems fine to put that in widgets.json if that makes things easier and cleaner than putting a call in a helper function.

Regarding changing the name of `properties`, yes, it would be less confusing to pick a different name. However, see next paragraph.

I like the mixin approach where the widgets.json file supplements or overrides the OAM file - I've been talking about it for awhile. But won't we ultimately need multiple operations: add an individual item to a field in the OAM, subtract an individual item to a field in the OAM, remove a field from the OAM (eg, remove all properties) and replace a field in the OAM (e.g., replace the list of properties). What I'm wondering is if it might make sense to wrap the widgets.json extensions to the OAM in some sort of command-like mechanism, such as:

```
"name":"MyWidget",
  ...
  oam:[
    {"target":"property","action":"add","name":"someproperty","value":true},
    {"target":"property","action":"remove","name":"someproperty"},
    {"target":"content","action":"replace","value":"<span></span>"}
  ]
```

I realize this is going overboard, but hear me out for just a second. Right now we have 2 <content> values, one in the OAM and one in the widgets.json. Maybe we could reconcile the 2 by having the <content> in the widgets.json be an override for the one in the OAM file. Also, the above approach would provide a way for the widgets.json file to remove properties from the OAM file that we don't feel are worthy for appearing in the widget-specific section. The model would be that OAM developers would create OAM properties for every possible property that the widget supports, and then tool-specific metadata files (e.g., our widgets.json file) could then decide which properties they actually wanted to present to the user.

**Comment by jhpedemonte** [comment]
> ...if that makes things easier and cleaner than putting a call in a helper function.

Yes, much better to support features **declaratively** rather than imperatively, when possible.

As for your counter-proposal, I'm a bit hesitant.  It seems a bit heavy.  I want to take some more time to think about it.

I liked my proposal because it was just a simple mixin.  However, as you point out, it won't let us remove entries from the OAM.

However, I think the _removal_ case will be quite rare.  The only situation that I am aware of (and which you mentioned) is to filter the number of listed properties.  But this could be handled by other methods.  [Previously](https://github.com/maqetta/maqetta/wiki/widgets.json-spec), I had proposed having a property whitelist or blacklist (or both), which would simply be an array of property names.  Much simpler that the syntax you outlined above.

If you can think of any other cases for having to _remove_ entries from the OAM, please mention.  Otherwise, I think it's acceptable to split out the functionality into separate concerns (a mixin and a whitelist/blacklist).

---

We could take my original proposal and expand it, such that the mixin is for the whole OAM, allowing us to mixin elements other than just `property`.

**Comment by jhpedemonte** [comment]
This last checkin fixed [use cases 2 and 4](https://github.com/maqetta/maqetta/issues/856#issuecomment-2205913).  Now, when the `fixed` property is changed, the VE reloads from source and the header/footer is in the correct location in DOM.

Still remaining is to fix the property not being read from source (`fixed` shows up as blank even if the property is set in source).

**Comment by jhpedemonte** [comment]
Also need to see if there is a way to restore the selection after the VE has been refreshed.

**Comment by JonFerraiolo** [comment]
Regarding restoring selection, one easy approach is to stash away the list of the IDs of all currently selected widgets in an array off on the side, and then reselect from that array after the VE has been refreshed. Using IDs in this manner would likely work 99% of the time, but I expect there will be  scenarios where the automatic ID logic will produce different automatic IDs during the refresh than what was there before the refresh. One possible hack would be to temporarily add an extra attribute to each widget on the HTML page with an auto-generated ID (e.g., maqtempid) that holds the ID to use during the refresh process, and then the setSourceData() logic (or whatever) knows to use that ID instead of creating a new one. And of course stripping that attribute out once processed.

**Comment by jhpedemonte** [comment]
I think this one is fixed.  I opened #1102 for restoring the widget selection after refresh. If there are any more issues related to this, just open a new bug.
